perllib/FixMyStreet/Queue/Item/Report.pm

```diff
@@ -202,39 +219,105 @@ sub _create_reporters {
     my $result = -1;
+    my $report = $self->report;
+
+    my @add_send_fail_body_ids;
+    my @remove_send_fail_body_ids;
+
+    for my $sender ( @{ $self->reporters } ) {
+        my $sender_name = ref $sender;
+        $self->log( 'Sending using ' . $sender_name );
+
+        if ( FixMyStreet->staging_flag( 'send_reports', 0 ) ) {
+            # If on staging with send_reports set to 0, bypass the body_id
+            # logic below, as we may only have a special Email reporter with
+            # no bodies attached
+            my $res = $sender->send( $report, $self->h );
+            $result *= $res;
+            $report->add_send_method($sender_name) if !$res;
+        }
+        elsif ( $sender_name =~ /${\EMAIL_SENDER_PREFIX}/ ) {
+            # Email is special-cased; an email to e.g. a two-tier council
+            # will have multiple recipients in the 'To' line. So we only
+            # need to record one failure (otherwise on retry/success,
+            # duplicate emails will be sent).
+            my $res = $sender->send( $self->report, $self->h );
+
+            $result *= $res;
+
+            my @body_ids = map { $_->id } @{ $sender->bodies };
+            if ($res) {
+                push @add_send_fail_body_ids, @body_ids;
+            }
+            else {
+                $report->add_send_method($sender_name);
+                push @remove_send_fail_body_ids, @body_ids;
+            }
+        }
+        else {
+            # Non-email senders.
+            # There can be multiple instances of the same type of sender, but
+            # each should only have one body.
+            for my $body ( @{ $sender->bodies } ) {
+                my $body_id = $body->id;
+
+                my $res = $sender->send( $self->report, $self->h );
+
+                $result *= $res;
+
+                if ($res) {
+                    push @add_send_fail_body_ids, $body_id;
+                }
+                else {
+                    $report->add_send_method($sender_name);
+                    push @remove_send_fail_body_ids, $body_id;
+                }
+            }
+        }
 
-    for my $sender ( keys %{$self->reporters} ) {
-        $self->log("Sending using " . $sender);
-        $sender = $self->reporters->{$sender};
-        my $res = $sender->send( $self->report, $self->h );
-        $result *= $res;
-        $self->report->add_send_method($sender) if !$res;
         if ( $self->manager ) {
-            if ($sender->unconfirmed_data) {
-                foreach my $e (keys %{ $sender->unconfirmed_data } ) {
-                    foreach my $c (keys %{ $sender->unconfirmed_data->{$e} }) {
-                        $self->manager->unconfirmed_data->{$e}{$c}{count} += $sender->unconfirmed_data->{$e}{$c}{count};
-                        $self->manager->unconfirmed_data->{$e}{$c}{note} = $sender->unconfirmed_data->{$e}{$c}{note};
+            if ( $sender->unconfirmed_data ) {
```

Comment:
This reformatting (automated?) is nothing to do with the fixup or the PR, I'd rather not have it if possible, thanks (in this particular case, I also don't think smaller multiple lines is clearer than slightly longer single lines, especially with `{count}` on its own line).